### PR TITLE
Support PR creation on Bitbucket

### LINKED ISSUE #2919: Add Bitbucket PR creation support
Bitbucket doesn't have an implementation for creating PRs. It would be great to have this, as I think this and #2918 are the only things keeping Bitbucket support from working in dependabot-script.

### PR DESCRIPTION
**Pull request by iinuwa**
This should fix #2919, but depends on #2918.

This gets us most of the way, but I'd appreciate some help polishing it off.

I was able to successfully create a PR with Bitbucket Cloud by modifying dependabot-script (cf. dependabot/dependabot-script#514), but I haven't added any automated tests because I don't really know where to start.

I also had a few questions as I was going through the code. I'll add review comments below.

### CODE REVIEW

File: common/lib/dependabot/clients/bitbucket.rb
Comment (iinuwa) [inline common/lib/dependabot/clients/bitbucket.rb:33]: Is this save to add an extra parameter here? (I hadn't done any Ruby before this, so I don't know about this.) It seems to match how the other constructors work.

File: common/lib/dependabot/clients/bitbucket.rb
Comment (iinuwa) [inline common/lib/dependabot/clients/bitbucket.rb:71]: I copied this most of the new methods from the Azure client, but I had to add the `repo` parameter to all of them.

Is this supposed to be using `source.repo` internally for all these parameters?

File: common/lib/dependabot/clients/bitbucket.rb
Comment (iinuwa) [inline common/lib/dependabot/clients/bitbucket.rb:73]: This only returns the last 100 commits; should this go paginate through all commits instead?

File: common/lib/dependabot/clients/bitbucket.rb
Comment (iinuwa) [inline common/lib/dependabot/clients/bitbucket.rb:101]: Are we supposed to bail if either `source_branch` or `target_branch` is `nil`? Or should we let the provider determine defaults for us if not specified in the client?

File: common/lib/dependabot/clients/bitbucket.rb
Comment (iinuwa) [inline common/lib/dependabot/clients/bitbucket.rb:148]: Bitbucket doesn't like the HTML tags in the generated description; does that need to be fixed in this PR before merging?

File: common/lib/dependabot/pull_request_creator/bitbucket.rb
Comment (iinuwa) [inline common/lib/dependabot/pull_request_creator/bitbucket.rb:35]: As the comment says, this logic needs to be reviewed.

File: common/lib/dependabot/pull_request_creator/bitbucket.rb
Comment (iinuwa) [inline common/lib/dependabot/pull_request_creator/bitbucket.rb:43]: I mentioned this above, but I'm not really sure what the `for_source` methods are for. Could someone explain this a little?

File: common/lib/dependabot/clients/bitbucket.rb
Comment (thepwagner) [reply common/lib/dependabot/clients/bitbucket.rb:73]: I suspect other clients do, but 100 commit seems like it might be "enough".
If it's working for you, works for me...

Comment (thepwagner) [inline common/lib/dependabot/clients/bitbucket.rb:73]:
```suggestion
        commits_path = "#{repo}/commits/#{branch_name}?pagelen=100"
```

Idea: simplify with string interpolation.

File: common/lib/dependabot/clients/bitbucket.rb
Comment (thepwagner) [inline common/lib/dependabot/clients/bitbucket.rb:122]: Concern: what value should `absolute_path` take if `file.name.start_with?("/") == true`?
It might be `nil` currently, which would mean multiple absolute paths conflict.

File: common/lib/dependabot/clients/bitbucket.rb
Comment (thepwagner) [inline common/lib/dependabot/clients/bitbucket.rb:128]: Xref https://developer.atlassian.com/bitbucket/api/2/reference/resource/repositories/%7Bworkspace%7D/%7Brepo_slug%7D/src#post

Not worth an inline comment, just sharing for fellow reviewers.

File: common/lib/dependabot/pull_request_creator/bitbucket.rb
Comment (thepwagner) [inline common/lib/dependabot/pull_request_creator/bitbucket.rb:87]:
```suggestion
          nil,
```

Comment SGTM, but since `labels` is a positional arg it needs to be provided so that `work_item` is passed correctly.

File: common/lib/dependabot/pull_request_creator/bitbucket.rb
Comment (thepwagner) [reply common/lib/dependabot/pull_request_creator/bitbucket.rb:43]: I think their main goal is to instantiate the client once and only once (e.g. the `@bitbucket_client_for_source ||=` line is the key).

I'm not sure about the naming - AFAICT there is  only ever one `source` (and therefore one `bitbucket_client`).

File: common/lib/dependabot/clients/bitbucket.rb
Comment (thepwagner) [reply common/lib/dependabot/clients/bitbucket.rb:33]: Question: what is using this new `source` field? I skimmed and don't see any references.

File: common/lib/dependabot/pull_request_creator/bitbucket.rb
Comment (thepwagner) [inline common/lib/dependabot/pull_request_creator/bitbucket.rb:52]: Nit: it's weird that we passed `source` to the constructor for `Dependabot::Clients::Bitbucket.for_source`, but still have to pass attributes like `source.repo` to individual methods.

If `source` isn't used by `Dependabot::Clients::Bitbucket` it may be cleanest to remove.

File: common/lib/dependabot/pull_request_creator/pr_name_prefixer.rb
Comment (thepwagner) [inline common/lib/dependabot/pull_request_creator/pr_name_prefixer.rb:312]:
```suggestion
```

I think this is expressed in the `commits()` function already.

File: common/lib/dependabot/pull_request_creator/pr_name_prefixer.rb
Comment (thepwagner) [inline common/lib/dependabot/pull_request_creator/pr_name_prefixer.rb:321]:
```suggestion
          map { |c| c.fetch("message") }.
          compact.
          reject { |m| m.start_with?("Merge") }.
```

Idea: reduce to commits with a `message` before filtering on the message contents.

File: common/lib/dependabot/clients/bitbucket.rb
Comment (thepwagner) [reply common/lib/dependabot/clients/bitbucket.rb:33]: It should be, since you're providing a default value.

**Review by thepwagner — COMMENTED** [review]
LGTM, thanks for this!

I'm also confused about how `source` provided to these clients' constructor should interact with `source.repo` required by methods. This long predates me, @feelepxyz or @hmarr is this part of an ongoing refactor? Which should be preferred?

File: common/lib/dependabot/clients/bitbucket.rb
Comment (iinuwa) [reply common/lib/dependabot/clients/bitbucket.rb]: Meh. I went ahead and implemented pagination for both commits and pull requests so someone else doesn't have to do it later.

File: common/lib/dependabot/clients/bitbucket.rb
Comment (iinuwa) [reply common/lib/dependabot/clients/bitbucket.rb]: Ah, that was a mistake; I've corrected that.

File: common/lib/dependabot/clients/bitbucket.rb
Comment (iinuwa) [reply common/lib/dependabot/clients/bitbucket.rb]: I didn't add any references to it since I wasn't sure whether I should prefer a `repo` parameter passed to each method, or to use the instance variable. I went ahead and removed the parameter since I don't think `source` applies to Bitbucket anyway.

(There isn't really an alternative location for Bitbucket. Support for Bitbucket Server is going away in Feb 2024 and the ability to configure custom domain names for Bitbucket Cloud isn't on the Atlassian roadmap to be implemented any time soon, so `bitbucket.org`/`api.bitbucket.org` are the only endpoints that we need to consider—unless someone who can test Bitbucket Server requests the feature.)

File: common/lib/dependabot/clients/bitbucket.rb
Comment (iinuwa) [reply common/lib/dependabot/clients/bitbucket.rb:19]: Yeah, I noticed that when I tried to remove the `for_source` method. It looks like all the clients do the same thing.

File: common/lib/dependabot/pull_request_creator/bitbucket.rb
Comment (iinuwa) [reply common/lib/dependabot/pull_request_creator/bitbucket.rb:35]: Updating PRs is a [different endpoint](https://developer.atlassian.com/bitbucket/api/2/reference/resource/repositories/%7Bworkspace%7D/%7Brepo_slug%7D/pullrequests/%7Bpull_request_id%7D#put), so we either need to throw an error when attempting to update, or add the endpoint reference.

File: common/lib/dependabot/pull_request_creator/bitbucket.rb
Comment (iinuwa) [reply common/lib/dependabot/pull_request_creator/bitbucket.rb]: Gotcha. I went ahead and removed the comment. I agree with the "only ever one source" part. 

> AFAICT there is only ever one `source`

Yeah, I agree. It seems to be the convention for the other clients, though, so I'll defer to you on the name.

File: common/lib/dependabot/pull_request_creator/pr_name_prefixer.rb
Comment (iinuwa) [reply common/lib/dependabot/pull_request_creator/pr_name_prefixer.rb]: Done, with a modification.

**Comment by iinuwa** [comment]
@thepwagner, I think I've addressed the suggestions you've made so far. I've put them in individual commits in case that's helpful for reviewing. I can squash and rebase after approval and before merging to clean up the history.

**Review by thepwagner — APPROVED** [review]
Thanks for this!
LGTM: I've regression tested with the `dry-run.rb` script in this repository.

dependabot/dependabot-script isn't supported by GitHub, so I haven't tested this there - follow-up PRs are welcome if more is needed.

**Comment by meladRaouf** [comment]
I can't use dependabot with bitcucket I got this error 
labeler.rb:241:in `labels': Unsupported provider bitbucket (RuntimeError)

is there anyway to skip labels in bitbucket

**Pull request merged**
